File: 2015/java/day03/Solution.java
package Day03;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class Solution {

    /*
 0 0 0 0 - 1
 1 0 0 0 - 2
 1 1 0 0 - 3
 1 1 1 0 - 4
 1 1 1 1 - 4

 0 0 - 1
 1 0 - 2
 1 1 - 3
 0 1 - 4
 0 0 - 4
*/
    private static String getInput() throws IOException {
        Path path = Paths.get("input.txt");
        return Files.readString(path);
    }

    private static int part1(String input) {
        int res = 1, ns = 0, ew = 0;

        List<int[]> list = new ArrayList<>();
        list.add(new int[]{ns, ew});

        for (int i = 0; i < input.length(); i++) {
            switch (input.charAt(i)) {
                case '^' -> ns++;
                case 'v' -> ns--;
                case '>' -> ew++;
                case '<' -> ew--;
            }
            int[] check = new int[]{ns, ew};

            boolean exists = false;
            for (int[] p : list) {
                if (p[0] == check[0] && p[1] == check[1]) {
                    exists = true;
                    break;
                }
            }

            if (!exists) {
                list.add(check);
                res++;
            }
        }
        return res;
    }

    private static int part2(String input) {
        int res = 1, ns = 0, ew = 0, rns = 0, rew = 0;

        List<int[]> list = new ArrayList<>();
        list.add(new int[]{ns, ew});

        for (int i = 0; i < input.length(); i++) {
            boolean robot = i%2 != 0;

            if (robot) {
                switch (input.charAt(i)) {
                    case '^' -> rns++;
                    case 'v' -> rns--;
                    case '>' -> rew++;
                    case '<' -> rew--;
                }
            } else {
                switch (input.charAt(i)) {
                    case '^' -> ns++;
                    case 'v' -> ns--;
                    case '>' -> ew++;
                    case '<' -> ew--;
                }
            }

            int[] check;

            if (robot) {
                check = new int[]{rns, rew};
            } else {
                check = new int[]{ns, ew};
            }

            boolean exists = false;
            for (int[] p : list) {
                if (p[0] == check[0] && p[1] == check[1]) {
                    exists = true;
                    break;
                }
            }

            if (!exists) {
                list.add(check);
                res++;
            }
        }

        return res;
    }

    public static void main(String[] args) {
        try {
            String input = getInput();
            System.out.println("Input : \n" + input);
            int res1 = part1(input);
            int res2 = part2(input);

            System.out.println("=== Final Results ===");
            System.out.printf("Part 1: %d%n", res1);
            System.out.printf("Part 2: %d%n", res2);
        } catch (IOException e) {
            System.err.println("Failed to read input: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: 2015/cpp/day03/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <utility>
#include <algorithm>
using namespace std;

string getInput() {
    ifstream file("input.txt");
    string input, line;

    while (getline(file, line)) {
        input += line + "\n";
    }

    file.close();
    return input;
}

void part1(string input) {
    int ns = 0, ew = 0;
    vector<pair<int, int>> houses = {make_pair(0, 0)};
    for (int i = 0; i < input.length(); i++) {
        switch (input[i])
        {
        case '^':
            ns++;
            break;
        case 'v':
            ns--;
            break;
        case '>':
            ew++;
            break;
        case '<':
            ew--;
            break;
        default:
            break;
        }

        auto tmp = find(houses.begin(), houses.end(), make_pair(ns, ew));
        if (tmp == houses.end()) {
            houses.push_back({ns, ew});
        }
    }

    cout << "Part1: " << houses.size() << endl;
}

void part2(string input) {
    int ns, ew, rns, rew;
    ns = ew = rns = rew = 0;
    vector<pair<int, int>> houses = {make_pair(0, 0)};
    vector<pair<int, int>> rhouses = {make_pair(0, 0)}; // Need to do -1 on the result calculation
    for (int i = 0; i < input.length(); i++) {
        bool robot = (i % 2 != 0);

        if (robot) {
            switch (input[i])
            {
            case '^':
                rns++;
                break;
            case 'v':
                rns--;
                break;
            case '>':
                rew++;
                break;
            case '<':
                rew--;
                break;
            default:
                break;
            }


            auto tmp = find(houses.begin(), houses.end(), make_pair(ns, ew));
            if (tmp == houses.end()) {
                houses.push_back({ns, ew});
            }
        } else {
            switch (input[i])
            {
            case '^':
                ns++;
                break;
            case 'v':
                ns--;
                break;
            case '>':
                ew++;
                break;
            case '<':
                ew--;
                break;
            default:
                break;
            }


            auto tmp = find(rhouses.begin(), rhouses.end(), make_pair(rns, rew));
            if (tmp == rhouses.end()) {
                rhouses.push_back({rns, rew});
            }
        }
    }

    int res = houses.size() + rhouses.size() - 1;
    cout << "Part1: " << res << endl;
}

int main() {
    string input = getInput();
    cout << "Input : \n" << input << endl;

    part1(input);
    part2(input);
    return 0;
}
